File: src/python/xerxes/session/branching.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Fork an existing session into a new one with the same history.

The child's ``parent_session_id`` points to the source so the lineage
is queryable. The metadata gets a ``forked_from`` annotation.

Exports:
    - branch_session
    - lineage"""

from __future__ import annotations

import copy
import uuid
from datetime import UTC, datetime

from .models import SessionRecord
from .store import SessionStore


def branch_session(
    store: SessionStore,
    *,
    source_session_id: str,
    title: str = "",
    new_session_id: str | None = None,
) -> SessionRecord:
    """Create a child session that shares the source's history.

    The new session inherits ``turns``, ``agent_transitions``, and
    ``metadata``; ``parent_session_id`` and timestamps are fresh."""
    source = store.load_session(source_session_id)
    if source is None:
        raise KeyError(f"unknown source session: {source_session_id}")
    now = datetime.now(UTC).isoformat()
    new = SessionRecord(
        session_id=new_session_id or uuid.uuid4().hex,
        workspace_id=source.workspace_id,
        created_at=now,
        updated_at=now,
        agent_id=source.agent_id,
        turns=copy.deepcopy(source.turns),
        agent_transitions=copy.deepcopy(source.agent_transitions),
        metadata={
            **source.metadata,
            "forked_from": source_session_id,
            "title": title or source.metadata.get("title", ""),
        },
        parent_session_id=source_session_id,
    )
    store.save_session(new)
    return new


def lineage(store: SessionStore, session_id: str) -> list[str]:
    """Return the chain of ``parent_session_id``s from this session to root.

    The first element is ``session_id``; the last has no parent."""
    chain: list[str] = []
    current = session_id
    while True:
        sess = store.load_session(current)
        if sess is None:
            break
        chain.append(sess.session_id)
        if not sess.parent_session_id:
            break
        current = sess.parent_session_id
    return chain


__all__ = ["branch_session", "lineage"]

File: xerxes/src/session/branching.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { randomUUID } from 'node:crypto'

import { cloneSessionRecord, SessionRecord } from './models.js'
import { isLockableSessionStore, type SessionStore } from './store.js'

export interface BranchSessionOptions {
  readonly newSessionId?: string
  readonly sourceSessionId: string
  readonly title?: string
}

/** Create a child session with a deep copy of the source conversation history. */
export function branchSession(store: SessionStore, options: BranchSessionOptions): SessionRecord {
  const source = store.loadSession(options.sourceSessionId)
  if (!source) throw new Error(`unknown source session: ${options.sourceSessionId}`)
  if (options.newSessionId !== undefined) {
    if (options.newSessionId === source.sessionId) {
      throw new Error('newSessionId must differ from sourceSessionId')
    }
    if (store.loadSession(options.newSessionId)) {
      throw new Error(`session already exists: ${options.newSessionId}`)
    }
  }

  const child = cloneSessionRecord(source)
  const now = new Date().toISOString()
  child.sessionId = options.newSessionId ?? randomUUID().replaceAll('-', '')
  child.createdAt = now
  child.updatedAt = now
  child.parentSessionId = source.sessionId
  child.metadata = {
    ...source.metadata,
    forked_from: source.sessionId,
    title: options.title || stringMetadata(source.metadata.title),
  }
  store.saveSession(child)
  return child
}

/**
 * {@link branchSession} with the whole check-then-create cycle performed under
 * the cross-process store lock. The synchronous form's existence checks and
 * final save are separate statements, so two processes branching to the same
 * explicit id could both pass the check and one silently overwrite the other;
 * under the shared lock exactly one wins and the other observes the created
 * row and fails its own check.
 */
export async function branchSessionExclusive(store: SessionStore, options: BranchSessionOptions): Promise<SessionRecord> {
  if (!isLockableSessionStore(store)) return branchSession(store, options)
  return store.withWriteLock(() => branchSession(store, options))
}

/** Follow parent pointers from a session back to its root, detecting cycles. */
export function sessionLineage(store: SessionStore, sessionId: string): string[] {
  const chain: string[] = []
  const seen = new Set<string>()
  let current: string | null = sessionId
  while (current && !seen.has(current)) {
    seen.add(current)
    const session = store.loadSession(current)
    if (!session) break
    chain.push(session.sessionId)
    current = session.parentSessionId
  }
  return chain
}

/** Compatibility alias for the original session API. */
export const lineage = sessionLineage

function stringMetadata(value: unknown): string {
  return typeof value === 'string' ? value : ''
}
